### PR TITLE
kubeadm: Remove cluster name from JoinConfiguration

### PR DESCRIPTION
**Pull request by ereslibre**
**What type of PR is this?**
/kind cleanup

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related https://github.com/kubernetes/kubeadm/issues/1088

**Special notes for your reviewer**:
None

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Review by timothysc — COMMENTED** [review]
Same comment about failed conversion test.

File: cmd/kubeadm/app/discovery/discovery.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func For(cfg *kubeadmapi.JoinConfiguration) (*clientcmdapi.Config, error) {
 	clusterinfo := kubeconfigutil.GetClusterFromKubeConfig(config)
 	return kubeconfigutil.CreateWithToken(
 		clusterinfo.Server,
-		cfg.ClusterName,
+		"cluster-discovery",
```
Comment (neolit123) [inline cmd/kubeadm/app/discovery/discovery.go]: shouldn't we be getting the clustername from the cluster here?

Comment (ereslibre) [reply cmd/kubeadm/app/discovery/discovery.go]: I don't see a way to do this at this point, do we want to call here `kubeconfigutil.ToClientSet` and then `configutil.FetchConfigFromFileOrCluster` in order to call `kubeconfigutil.CreateWithToken` with the correct cluster name?

**Comment by ereslibre** [comment]
I will add the conversion test tomorrow.

File: cmd/kubeadm/app/discovery/discovery.go (hunk 1 above)
Comment (ereslibre) [reply cmd/kubeadm/app/discovery/discovery.go]: @neolit123 I have added a separate commit addressing your comment, fetching the cluster name from the context, I don't know if it's the best approach.

**Review by fabriziopandini — COMMENTED** [review]
@ereslibre thanks for this PR!

if I got this right, `JoinConfiguration.ClusterName` is used only to set the cluster name in the config map that gets returned from Discovery, and you are proposing to replace it with a constant.

I'm fine with this, even if I would use existing const `DefaultClusterName` instead of creating a new const & I will remove parameter passing this const along the chain of calls.

Now, what should be decided is if it is fine to keep the above constant as a clusterName when the above config will be saved into the `bootstrap-kubelet.conf` or if we want to replace the cluster name with ClusterConfiguration.ClusterName before saving `bootstrap-kubelet.conf`.

In other words:

Discovery --> TLSBootstrap kubeconfig with clusterName = Constants --> bootstrap-kubelet.conf

or 

Discovery --> TLSBootstrap kubeconfig with clusterName = Constants --> Replace cluster name with ClusterConfiguration.ClusterName --> bootstrap-kubelet.conf

@ereslibre @luxas @timothysc @neolit123 opinions?

**Comment by neolit123** [comment]
> Discovery --> TLSBootstrap kubeconfig with clusterName = Constants --> Replace cluster name with ClusterConfiguration.ClusterName --> bootstrap-kubelet.conf

+1 for this.
no strong opinions, but if go for the other option it should be `DefaultClusterName` at least.

**Comment by ereslibre** [comment]
> Discovery --> TLSBootstrap kubeconfig with clusterName = Constants --> Replace cluster name with ClusterConfiguration.ClusterName --> bootstrap-kubelet.conf

+1 for this, will update the PR today.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go [hunk 2]
```diff
@@ -28,6 +29,10 @@ func Convert_v1alpha3_JoinConfiguration_To_kubeadm_JoinConfiguration(in *JoinCon
 		return err
 	}
 
+	if len(in.ClusterName) != 0 {
+		return errors.New("clusterName has been removed from JoinConfiguration and clusterName from ClusterConfiguration will be used instead. Please cleanup JoinConfiguration.ClusterName fields")
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go]: i don't think this is needed, but will defer to @fabriziopandini for second opinion.
also PSA that the "global" yaml parser we are using will "soon" going to start throw errors on unknown and duplicate fields. hopefully in 1.13.

File: cmd/kubeadm/app/cmd/join.go [hunk 3]
```diff
@@ -607,7 +607,7 @@ func fetchInitConfigurationFromJoinConfiguration(cfg *kubeadmapi.JoinConfigurati
 	// Perform the Discovery, which turns a Bootstrap Token and optionally (and preferably) a CA cert hash into a KubeConfig
 	// file that may be used for the TLS Bootstrapping process the kubelet performs using the Certificates API.
 	glog.V(1).Infoln("[join] Discovering cluster-info")
-	tlsBootstrapCfg, err := discovery.For(cfg)
+	tlsBootstrapCfg, err := discovery.For(cfg, kubeadmconstants.DefaultClusterName)
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: what if the already existing control-plane node is not using the default cluster name?

File: cmd/kubeadm/app/constants/constants.go [hunk 4]
```diff
@@ -178,6 +178,9 @@ const (
 	// TLSBootstrapTimeout specifies how long kubeadm should wait for the kubelet to perform the TLS Bootstrap
 	TLSBootstrapTimeout = 2 * time.Minute
 
+	// DefaultClusterName defines the default cluster name
+	DefaultClusterName = "kubernetes"
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go]: this change makes sense to me.
i.e. removing the constant from the versions / defaults.go files.

File: cmd/kubeadm/app/cmd/join.go (hunk 3 above)
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/join.go]: This is the first step to fetch the `ClusterConfiguration`, not sure I understand the question. Right after this and after fetching the `ClusterConfiguration` with this temporary kubeconfig the final one will be set here: https://github.com/kubernetes/kubernetes/pull/70757/files#diff-0960dc0bb7c3e7113a0daa027856b8f9R623

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go]: I am ok with this

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults.go [hunk 5]
```diff
@@ -42,9 +42,6 @@ const (
 	DefaultImageRepository = "k8s.gcr.io"
 	// DefaultManifestsDir defines default manifests directory
 	DefaultManifestsDir = "/etc/kubernetes/manifests"
-	// DefaultClusterName defines the default cluster name
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults.go]: I don't think we should move this cost alone because we end in a mixed situation where some API defaults are defined here on some in another place

File: cmd/kubeadm/app/cmd/join.go [hunk 6]
```diff
@@ -619,6 +619,12 @@ func fetchInitConfigurationFromJoinConfiguration(cfg *kubeadmapi.JoinConfigurati
 		return nil, nil, err
 	}
 
+	// Retrieve the final KubeConfig file with the cluster name discovered after fetching the cluster configuration
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/join.go]: I don't think we should do discovery twice.
Instead I think we should change the TLS kubeconfig in place or create a new kubeconfig file from TLS kubeconfig but with a different cluster name (if you need prior art let me know)

This will allow to clean up the Discovery.For chain of calls from the clusterName field

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@ereslibre 
This version introduced a double second discovery, that is something we should avoid.
Could you kindly check if the approach I'm suggesting in comments is feasible for you?

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults.go (hunk 5 above)
Comment (ereslibre) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults.go]: I'm restoring this and using `v1beta1` package instead for now.

File: cmd/kubeadm/app/cmd/join.go [hunk 7]
```diff
@@ -619,6 +619,16 @@ func fetchInitConfigurationFromJoinConfiguration(cfg *kubeadmapi.JoinConfigurati
 		return nil, nil, err
 	}
 
+	// Create the final KubeConfig file with the cluster name discovered after fetching the cluster configuration
+	clusterinfo := kubeconfigutil.GetClusterFromKubeConfig(tlsBootstrapCfg)
+	tlsBootstrapCfg = kubeconfigutil.CreateWithToken(
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: This doesn't work, as the client cert returned from TLS bootstrap would be overridden with the tls bootstrap token.
The way to do this is:
1. Extract `clusterinfo` (like you do)
2. Delete the existing cluster key and value in the .Clusters map
3. Replace the current context's `.Cluster` value
4. Add `clusterinfo` back to the `.Clusters` map with the new `initConfiguration.ClusterName` key

**Review by luxas — COMMENTED** [review]
LGTM apart from the comment

File: cmd/kubeadm/app/cmd/join.go (hunk 7 above)
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/join.go]: @luxas Out of curiosity, in the case you are mentioning, this is where `discovery.For` is supposed to return? (so `len(cfg.Discovery.TLSBootstrapToken) == 0`): https://github.com/kubernetes/kubernetes/blob/91d6d7530382bf120a1575bae72ca36a33842e95/cmd/kubeadm/app/discovery/discovery.go#L45-L47

**Comment by ereslibre** [comment]
PR rebased.

**Review by timothysc — APPROVED** [review]
/approve
